core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -239,9 +239,13 @@ class GroupMetadataManager(brokerId: Int,
     }
   }
 
+  /**
+   * @return Returning a map of successfully appended topic partitions and a flag indicting whether the HWM has been
+   *         incremented. The caller ought to complete delayed requests for those returned partitions.
+   */
   def storeGroup(group: GroupMetadata,
                  groupAssignment: Map[String, Array[Byte]],
-                 responseCallback: Errors => Unit): Unit = {
+                 responseCallback: Errors => Unit): Map[TopicPartition, LeaderHWChange] = {
```

Comment:
It's weird to have a method that invokes a callback and returns a result. Do we need both?

Reply:
@ijuma thanks for your reviews.

>  It would be good to reconsider that as it's difficult to reason about usage in such cases.

You are totally right and @hachikuji had given a great refactor idea (https://github.com/apache/kafka/pull/8657#discussion_r426943627). Given that refactor will bring a lot of changes to this PR, I had filed a ticket to refactor related code (see https://issues.apache.org/jira/browse/KAFKA-10170) in order to make this PR focus on bug fix.